clients/go/admin/token_source_provider.go

```diff
@@ -141,15 +147,68 @@ func NewClientCredentialsTokenSourceProvider(ctx context.Context, cfg *Config,
 	if len(scopes) == 0 {
 		scopes = clientMetadata.Scopes
 	}
-
 	return ClientCredentialsTokenSourceProvider{
 		ccConfig: clientcredentials.Config{
 			ClientID:     cfg.ClientID,
 			ClientSecret: secret,
 			TokenURL:     tokenURL,
-			Scopes:       scopes}}, nil
+			Scopes:       scopes},
+		MaxDurationBeforeTokenExpiry: cfg.PreemptiveRefreshDuration.Duration,
+		RefreshTokenPreemptively:     cfg.RefreshTokenPreemptively}, nil
 }
 
 func (p ClientCredentialsTokenSourceProvider) GetTokenSource(ctx context.Context) (oauth2.TokenSource, error) {
+	if p.RefreshTokenPreemptively {
+		source := p.ccConfig.TokenSource(ctx)
+		return &customTokenSource{
+			new:                          source,
+			mu:                           sync.Mutex{},
+			t:                            nil,
+			maxDurationBeforeTokenExpiry: p.MaxDurationBeforeTokenExpiry,
+		}, nil
+	}
 	return p.ccConfig.TokenSource(ctx), nil
 }
+
+type customTokenSource struct {
+	new                          oauth2.TokenSource
+	mu                           sync.Mutex // guards everything else
+	t                            *oauth2.Token
+	refreshTime                  time.Time
+	refreshed                    bool
```

Comment:
This is really `failedToRefresh`. Right?

Reply:
Yes, let me rename it.